File: apps/ai-engine/services/graph_loader.py
"""
Supabase → NetworkX DiGraph 로드
TypeScript graph-loader.ts와 동일한 로직 포팅

반환: (graph: nx.DiGraph, unit_labels: dict[uuid, label_str])
  - 노드 ID = DB uuid (mcs_equipment_unit.id)
  - 노드 속성: label (equipment_unit_id), unit_type, in_out_mode
  - 엣지 속성: weight
"""
import networkx as nx
from typing import Tuple, Dict
from services.supabase_client import get_supabase


def load_graph(layout_id: str) -> Tuple[nx.DiGraph, Dict[str, str]]:
    """
    특정 레이아웃의 전이 관계를 Supabase에서 로드하여 방향 가중 그래프 구성

    Args:
        layout_id: mcs_layout.id (uuid)

    Returns:
        (graph, unit_labels)
        - graph: NetworkX DiGraph (노드=unit_uuid, 엣지=departure→arrival+weight)
        - unit_labels: {unit_uuid: equipment_unit_id 라벨}

    Raises:
        ValueError: 레이아웃에 전이 관계가 없는 경우
    """
    client = get_supabase()

    # 1단계: 해당 레이아웃의 장비 ID 목록 조회
    resp = client.table("mcs_equipment").select("id").eq("layout_id", layout_id).execute()
    equipments = resp.data or []
    if not equipments:
        raise ValueError(f"레이아웃({layout_id})에 장비가 없습니다. 레이아웃을 먼저 저장해주세요.")

    equipment_ids = [e["id"] for e in equipments]

    # 2단계: 해당 장비들의 유닛 목록 조회
    units_resp = (
        client.table("mcs_equipment_unit")
        .select("id, equipment_unit_id, unit_type, in_out_mode")
        .in_("equipment_id", equipment_ids)
        .execute()
    )
    units = units_resp.data or []

    # 3단계: 전이 관계 로드
    rel_resp = (
        client.table("mcs_transfer_relation")
        .select("departure_unit_id, arrival_unit_id, weight")
        .eq("layout_id", layout_id)
        .execute()
    )
    relations = rel_resp.data or []

    if not relations:
        # 폴백: 레이아웃 JSON에서 직접 그래프 빌드 + DB 자동 복구
        print(f"[graph_loader] 전이 관계 없음 → 레이아웃 JSON 폴백 시도 (layout_id={layout_id})")
        relations = _build_relations_from_json(layout_id, client, units)
        if not relations:
            raise ValueError(
                f"레이아웃({layout_id})에 전이 관계가 없습니다. 레이아웃 모델러에서 저장 후 사용하세요."
            )

    # 유닛 정보 인덱싱
    unit_map: Dict[str, dict] = {u["id"]: u for u in units}
    unit_labels: Dict[str, str] = {u["id"]: u["equipment_unit_id"] for u in units}

    # 4단계: NetworkX 방향 가중 그래프 구성
    graph = nx.DiGraph()

    # 노드 추가 (전이 관계에 등장하는 유닛만)
    for rel in relations:
        for unit_id in [rel["departure_unit_id"], rel["arrival_unit_id"]]:
            if unit_id not in graph and unit_id in unit_map:
                u = unit_map[unit_id]
                graph.add_node(
                    unit_id,
                    label=u["equipment_unit_id"],
                    unit_type=u["unit_type"],
                    in_out_mode=u["in_out_mode"],
                )

    # 엣지 추가 (방향 가중)
    for rel in relations:
        dep = rel["departure_unit_id"]
        arr = rel["arrival_unit_id"]
        weight = float(rel["weight"])
        if dep in graph and arr in graph:
            graph.add_edge(dep, arr, weight=weight)

    return graph, unit_labels


def _build_relations_from_json(
    layout_id: str,
    client,
    units: list[dict],
) -> list[dict]:
    """
    레이아웃 JSON(json_data)에서 전이 관계를 재구성하여 반환하고, DB에도 자동 복구 삽입.

    syncLayoutToDb 가 실패했거나 오래된 저장본에 전이 관계가 없는 경우 폴백으로 사용.
    """
    # 레이아웃 JSON 조회
    layout_resp = (
        client.table("mcs_layout")
        .select("json_data")
        .eq("id", layout_id)
        .maybe_single()
        .execute()
    )
    if not layout_resp.data:
        return []

    json_data = layout_resp.data.get("json_data") or {}
    nodes = json_data.get("nodes", [])
    edges = json_data.get("edges", [])

    # 유닛 코드 → DB uuid 인덱스 (equipment_unit_id → id)
    units_by_code: dict[str, str] = {u["equipment_unit_id"]: u["id"] for u in units}

    # React Flow node.id → unit DB uuid 매핑 구성
    rf_to_unit_id: dict[str, str] = {}
    for node in nodes:
        node_id = node.get("id", "")
        node_type = node.get("type", "")
        data = node.get("data") or {}

        if node_type == "port":
            unit_code = (data.get("portId") or "").strip() or node_id
        elif node_type in ("node", "charge"):
            # charge 노드는 nodeId 필드 대신 nodeId 사용 (layout-modeler 규칙 동일)
            unit_code = (data.get("nodeId") or "").strip() or node_id
        else:
            # stocker/process/agv 는 port/node 유닛으로만 접근
            continue

        db_id = units_by_code.get(unit_code)
        if db_id:
            rf_to_unit_id[node_id] = db_id

    # transfer 엣지에서 전이 관계 구성
    relations = []
    for edge in edges:
        if edge.get("type") != "transfer":
            continue
        source = edge.get("source", "")
        target = edge.get("target", "")
        edge_data = edge.get("data") or {}
        weight = float(edge_data.get("weight") or 1.0)

        dep_id = rf_to_unit_id.get(source)
        arr_id = rf_to_unit_id.get(target)
        if not dep_id or not arr_id:
            continue

        relations.append({
            "layout_id":         layout_id,
            "departure_unit_id": dep_id,
            "arrival_unit_id":   arr_id,
            "weight":            weight,
        })

    if not relations:
        print(f"[graph_loader] JSON 폴백에서도 전이 관계를 찾을 수 없음 (엣지 수: {len(edges)})")
        return []

    # DB에 자동 복구 삽입 (이후 호출 시 폴백 없이 바로 로드 가능)
    try:
        client.table("mcs_transfer_relation").insert(relations).execute()
        print(f"[graph_loader] 전이 관계 {len(relations)}건 자동 복구 삽입 완료")
    except Exception as e:
        # 삽입 실패해도 이번 호출은 relations 메모리 데이터로 계속 진행
        print(f"[graph_loader] 자동 복구 삽입 실패 (무시): {e}")

    return [
        {"departure_unit_id": r["departure_unit_id"], "arrival_unit_id": r["arrival_unit_id"], "weight": r["weight"]}
        for r in relations
    ]

File: apps/mcs/lib/engine/graph-loader.ts
import type { SupabaseClient } from '@supabase/supabase-js';
import { createAdminClient } from '@/lib/supabase/server';
import type { GraphNode } from './types';

/**
 * Supabase에서 특정 레이아웃의 전이 관계를 로드하여 A* 그래프 구성
 *
 * mcs_transfer_relation: departure_unit_id → arrival_unit_id, weight
 * mcs_equipment_unit:    id, equipment_unit_id (표시용 라벨)
 *
 * 전이 관계가 없으면 레이아웃 JSON에서 직접 복구 후 DB에 자동 저장.
 *
 * @returns unitId(DB uuid) → GraphNode 맵
 */
export async function loadGraph(layoutId: string): Promise<Map<string, GraphNode>> {
  // RLS 우회: API 라우트에서 세션 없이 호출되므로 service_role 클라이언트 사용
  const supabase = createAdminClient();

  // 전이 관계 전체 로드
  const { data: relations, error: relErr } = await supabase
    .from('mcs_transfer_relation')
    .select('departure_unit_id, arrival_unit_id, weight')
    .eq('layout_id', layoutId);
  if (relErr) throw relErr;

  // 전이 관계가 없으면 레이아웃 JSON에서 복구 시도
  const effectiveRelations = (relations && relations.length > 0)
    ? relations
    : await recoverRelationsFromJson(layoutId, supabase);

  if (!effectiveRelations || effectiveRelations.length === 0) {
    throw new Error(`레이아웃(${layoutId})에 전이 관계가 없습니다. 레이아웃 모델러에서 저장 후 사용하세요.`);
  }

  // 관련 유닛 ID 수집
  const unitIds = Array.from(
    new Set(effectiveRelations.flatMap((r) => [r.departure_unit_id, r.arrival_unit_id]))
  );

  // 유닛 표시용 라벨 로드
  const { data: units, error: unitErr } = await supabase
    .from('mcs_equipment_unit')
    .select('id, equipment_unit_id')
    .in('id', unitIds);
  if (unitErr) throw unitErr;

  const unitLabelMap = new Map<string, string>();
  (units ?? []).forEach((u) => unitLabelMap.set(u.id, u.equipment_unit_id));

  // 그래프 맵 구성
  const graph = new Map<string, GraphNode>();

  // 모든 유닛 노드 초기화
  for (const unitId of unitIds) {
    graph.set(unitId, {
      id:        unitId,
      unitId:    unitLabelMap.get(unitId) ?? unitId,
      neighbors: [],
    });
  }

  // 엣지 추가 (방향 그래프)
  for (const rel of effectiveRelations) {
    const node = graph.get(rel.departure_unit_id);
    if (node) {
      node.neighbors.push({
        toUnitId: rel.arrival_unit_id,
        weight:   Number(rel.weight),
      });
    }
  }

  return graph;
}

/** 레이아웃 JSON에서 전이 관계를 재구성하여 DB에 자동 복구 삽입 후 반환 */
async function recoverRelationsFromJson(
  layoutId: string,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  supabase: SupabaseClient<any>,
): Promise<Array<{ departure_unit_id: string; arrival_unit_id: string; weight: number }>> {
  console.warn(`[graph-loader] 전이 관계 없음 → JSON 폴백 복구 시도 (layoutId=${layoutId})`);

  // 레이아웃 JSON 조회
  const { data: layout } = await supabase
    .from('mcs_layout')
    .select('json_data')
    .eq('id', layoutId)
    .maybeSingle();
  if (!layout?.json_data) return [];

  const jsonData = layout.json_data as { nodes?: unknown[]; edges?: unknown[] };
  const nodes = (jsonData.nodes ?? []) as Array<{ id: string; type?: string; data?: Record<string, unknown> }>;
  const edges = (jsonData.edges ?? []) as Array<{ source: string; target: string; type?: string; data?: Record<string, unknown> }>;

  // 레이아웃의 모든 유닛 조회 (equipment_unit_id → DB uuid 매핑용)
  const { data: equipments } = await supabase
    .from('mcs_equipment')
    .select('id')
    .eq('layout_id', layoutId);
  const equipmentIds = (equipments ?? []).map((e: { id: string }) => e.id);
  if (equipmentIds.length === 0) return [];

  const { data: units } = await supabase
    .from('mcs_equipment_unit')
    .select('id, equipment_unit_id')
    .in('equipment_id', equipmentIds);

  const unitsByCode = new Map<string, string>();
  (units ?? []).forEach((u: { id: string; equipment_unit_id: string }) => {
    unitsByCode.set(u.equipment_unit_id, u.id);
  });

  // React Flow node.id → unit DB uuid 매핑
  const rfToUnitId = new Map<string, string>();
  for (const node of nodes) {
    const { id: rfId, type: nodeType, data = {} } = node;
    let unitCode: string | undefined;
    if (nodeType === 'port') {
      unitCode = ((data.portId as string | undefined) ?? '').trim() || rfId;
    } else if (nodeType === 'node' || nodeType === 'charge') {
      unitCode = ((data.nodeId as string | undefined) ?? '').trim() || rfId;
    } else {
      continue;
    }
    const dbId = unitsByCode.get(unitCode);
    if (dbId) rfToUnitId.set(rfId, dbId);
  }

  // transfer 엣지 → 전이 관계 목록
  const relations: Array<{ layout_id: string; departure_unit_id: string; arrival_unit_id: string; weight: number }> = [];
  for (const edge of edges) {
    if (edge.type !== 'transfer') continue;
    const depId = rfToUnitId.get(edge.source);
    const arrId = rfToUnitId.get(edge.target);
    if (!depId || !arrId) continue;
    const weight = Number((edge.data?.weight as number | undefined) ?? 1.0);
    relations.push({ layout_id: layoutId, departure_unit_id: depId, arrival_unit_id: arrId, weight });
  }

  if (relations.length === 0) {
    console.warn(`[graph-loader] JSON 폴백에서도 전이 관계를 찾을 수 없음 (엣지 수: ${edges.length})`);
    return [];
  }

  // DB에 자동 복구 삽입 (이후 호출 시 폴백 없이 바로 로드)
  try {
    await supabase.from('mcs_transfer_relation').insert(relations);
    console.info(`[graph-loader] 전이 관계 ${relations.length}건 자동 복구 완료`);
  } catch (e) {
    console.warn('[graph-loader] 자동 복구 삽입 실패 (무시):', e);
  }

  return relations.map(({ departure_unit_id, arrival_unit_id, weight }) => ({
    departure_unit_id,
    arrival_unit_id,
    weight,
  }));
}
